stage1/init/common/pod.go

```diff
@@ -412,48 +413,27 @@ func parseUserGroup(p *stage1commontypes.Pod, ra *schema.RuntimeApp) (int, int,
 // It returns the actual path relative to the app rootfs for the given path.
 // This is needed for absolute symlinks - we are in a different rootfs.
 func EvaluateSymlinksInsideApp(appRootfs, path string) (string, error) {
-	link := appRootfs
-
-	paths := strings.Split(path, "/")
-	for i, p := range paths {
-		next := filepath.Join(link, p)
-
-		if !strings.HasPrefix(next, appRootfs) {
-			return "", fmt.Errorf("path escapes app's root: %q", path)
-		}
-
-		fi, err := os.Lstat(next)
-		if err != nil {
-			if os.IsNotExist(err) {
-				link = filepath.Join(append([]string{link}, paths[i:]...)...)
-				break
-			}
-			return "", err
-		}
-
-		if fi.Mode()&os.ModeType != os.ModeSymlink {
-			link = filepath.Join(link, p)
-			continue
-		}
+	chroot, err := newChroot(appRootfs)
+	if err != nil {
+		return "", errwrap.Wrapf(fmt.Sprintf("chroot to %q failed", appRootfs), err)
+	}
 
-		// Evaluate the symlink.
-		target, err := os.Readlink(next)
-		if err != nil {
-			return "", err
-		}
+	target, err := fileutil.EvalSymlinksAlways(path)
+	if err != nil {
+		return "", errwrap.Wrapf(fmt.Sprintf("evaluating symlinks of %q failed", path), err)
+	}
 
-		if filepath.IsAbs(target) {
-			link = filepath.Join(appRootfs, target)
-		} else {
-			link = filepath.Join(link, target)
-		}
+	// EvalSymlinksAlways might return a relative path
+	abs, err := filepath.Abs(target)
+	if err != nil {
+		return "", errwrap.Wrapf(fmt.Sprintf("failed to get absolute representation of %q", target), err)
+	}
 
-		if !strings.HasPrefix(link, appRootfs) {
-			return "", fmt.Errorf("symlink %q escapes app's root with value %q", next, target)
-		}
+	if err := chroot.escape(); err != nil {
```

Comment:
If I move this right after `EvalSymlinksAlways`, then `filepath.Abs` will be invoked in the outer rootfs.